Convert crm_resource to formatted output

clumens opened the pull request:
This isn't quite ready yet.  First, see the WIP commit.  Second, a couple of places are getting this:
```
[root@cluster01 ~]# crm_resource -C --output-as xml
Waiting for 2 replies from the controller.. OK
<pacemaker-result api-version="2.3" request="crm_resource -C --output-as xml">
  <status code="0" message="OK"/>
</pacemaker-result>
```

And third, a couple of places are getting this:
```
[root@cluster01 ~]# crm_resource --move -r dummy --output-as xml
<pacemaker-result api-version="2.3" request="crm_resource --move -r dummy --output-as xml">
  <status code="0" message="OK">
    <errors>
      <error>WARNING: Creating rsc_location constraint 'cli-ban-dummy-on-cluster02' with a score of -INFINITY for resource dummy on cluster02.
        This will prevent dummy from running on cluster02 until the constraint is removed using the clear option or by editing the CIB with an appropriate tool
        This will be the case even if cluster02 is the last node in the cluster</error>
    </errors>
  </status>
</pacemaker-result>
```

But I'd like to start talking about the rest of it.  My intention is to get this in after the next release.  I don't think it will have had nearly enough testing by then.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -221,7 +221,8 @@ find_matching_attr_resources(pcmk__output_t *out, pe_resource_t * rsc,
             if (!out->is_quiet(out)) {
-                printf("Performing %s of '%s' on '%s', the parent of '%s'\n", cmd, attr_name, rsc->id, rsc_id);
+                out->info(out, "Performing %s of '%s' on '%s', the parent of '%s'",
+                          cmd, attr_name, rsc->id, rsc_id);
```
Knowing that a different resource was used (rsc->id) is potentially useful to have in XML. It looks like extracting that would be a pain though, so I wouldn't worry about it unless demand arises in the future

kgaillot commented on cts/cli/regression.tools.exp:
```diff
@@ -2467,7 +2467,6 @@ Set 'test-primitive' option: id=test-primitive-meta_attributes-is-managed set=te
 Multiple attributes match name=is-managed
   Value: false 	(id=test-primitive-meta_attributes-is-managed)
   Value: false 	(id=test-clone-meta_attributes-is-managed)
-
```
I wonder if it would be useful to have an out->spacer() method (newline in text, nothing in xml, maybe a line break or horizontal line in html)

clumens replied:
I think we've talked about that before.  I'm using out->info with an empty string to accomplish that right now.  See PCMK__OUTPUT_SPACE_IF.  The fact that it's come up twice makes me think it's a good idea.

clumens replied:
https://github.com/ClusterLabs/pacemaker/pull/2198

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -283,10 +283,10 @@ cli_resource_update_attribute(pcmk__output_t *out, pe_resource_t *rsc,
+                out->info(out, "WARNING: There is already a meta attribute for '%s' called '%s' (id=%s)",
+                          uber_parent(rsc)->id, attr_name, local_attr_id);
+                out->info(out, "         Delete '%s' first or use the force option to override",
+                          local_attr_id);
```
I'm thinking maybe this should be ->err() since we're returning an error code

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1797,7 +1789,7 @@ cli_resource_execute_from_params(pcmk__output_t *out, const char *rsc_name,
                 if (local_copy[lpc] == '\n' || local_copy[lpc] == 0) {
                     local_copy[lpc] = 0;
-                    printf(" >  stdout: %s\n", local_copy + last);
+                    out->info(out, " >  stdout: %s", local_copy + last);
```
I think this would be useful in XML, like ->subprocess_output(). Not sure if that could just be used directly

clumens replied:
Sure, subprocess_output just takes strings.  There's no reason we couldn't use that here.

kgaillot commented on lib/pengine/pe_output.c:
```diff
@@ -1284,6 +1284,109 @@ pe__node_attribute_html(pcmk__output_t *out, va_list args) {
 
+PCMK__OUTPUT_ARGS("node-and-op", "pe_working_set_t *", "xmlNode *")
+int
+pe__node_and_op(pcmk__output_t *out, va_list args) {
```
What are the differences from op-history? Maybe this is one of the things worth changing for 2.1.0

clumens replied:
This would be a great candidate to merge into something else.  I would really kind of like to just drop the node-and-op message and reuse the stuff from crm_mon, if we could get away with it.  I introduced this to preserve existing output.

node-and-op output looks like this:
```
[root@cluster01 ~]# crm_resource --list-operations
ping    (ocf::pacemaker:ping):   Started: ping_start_0 (node=cluster02, call=16, rc=0, last-rc-change=Mon Oct 12 14:38:13 2020, exec=2053ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_monitor_10000 (node=cluster02, call=17, rc=0, last-rc-change=Mon Oct 12 14:38:15 2020, exec=2039ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_monitor_0 (node=cluster02, call=10, rc=7, last-rc-change=Mon Oct 12 14:38:03 2020, exec=3ms): complete
dummy   (ocf::pacemaker:Dummy):  Started: dummy_start_0 (node=cluster02, call=18, rc=0, last-rc-change=Mon Oct 12 14:38:19 2020, exec=6019ms): complete
dummy   (ocf::pacemaker:Dummy):  Started: dummy_monitor_60000 (node=cluster02, call=19, rc=0, last-rc-change=Mon Oct 12 14:38:26 2020, exec=6015ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_start_0 (node=cluster01, call=11, rc=0, last-rc-change=Mon Oct 12 14:39:52 2020, exec=2095ms): complete
ping    (ocf::pacemaker:ping):   Started: ping_monitor_10000 (node=cluster01, call=12, rc=0, last-rc-change=Mon Oct 12 14:39:54 2020, exec=2071ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_start_0 (node=cluster01, call=13, rc=0, last-rc-change=Mon Oct 12 14:39:56 2020, exec=146ms): complete
Fencing (stonith:fence_xvm):     Started: Fencing_monitor_60000 (node=cluster01, call=14, rc=0, last-rc-change=Mon Oct 12 14:39:56 2020, exec=75ms): complete
```

And op-history output looks like this:
```
[root@cluster01 ~]# crm_mon --include=none,operations -1t
Operations:
  * Node: cluster02:
    * ping: migration-threshold=1000000:
      * (16) start: last-rc-change="Mon Oct 12 14:38:13 2020" last-run="Mon Oct 12 14:38:13 2020" exec-time="2053ms" queue-time="0ms" rc=0 (ok)
      * (17) monitor: interval="10000ms" last-rc-change="Mon Oct 12 14:38:15 2020" exec-time="2039ms" queue-time="0ms" rc=0 (ok)
    * dummy: migration-threshold=1000000:
      * (14) probe: last-rc-change="Mon Oct 12 14:38:03 2020" last-run="Mon Oct 12 14:38:03 2020" exec-time="5002ms" queue-time="0ms" rc=1 (error)
      * (18) start: last-rc-change="Mon Oct 12 14:38:19 2020" last-run="Mon Oct 12 14:38:19 2020" exec-time="6019ms" queue-time="0ms" rc=0 (ok)
      * (19) monitor: interval="60000ms" last-rc-change="Mon Oct 12 14:38:26 2020" exec-time="6015ms" queue-time="0ms" rc=0 (ok)
  * Node: cluster01:
    * ping: migration-threshold=1000000:
      * (11) start: last-rc-change="Mon Oct 12 14:39:52 2020" last-run="Mon Oct 12 14:39:52 2020" exec-time="2095ms" queue-time="0ms" rc=0 (ok)
      * (12) monitor: interval="10000ms" last-rc-change="Mon Oct 12 14:39:54 2020" exec-time="2071ms" queue-time="0ms" rc=0 (ok)
    * Fencing: migration-threshold=1000000:
      * (13) start: last-rc-change="Mon Oct 12 14:39:56 2020" last-run="Mon Oct 12 14:39:56 2020" exec-time="146ms" queue-time="0ms" rc=0 (ok)
      * (14) monitor: interval="60000ms" last-rc-change="Mon Oct 12 14:39:56 2020" exec-time="75ms" queue-time="0ms" rc=0 (ok)
    * dummy: migration-threshold=1000000:
      * (18) probe: last-rc-change="Mon Oct 12 14:39:57 2020" last-run="Mon Oct 12 14:39:57 2020" exec-time="5043ms" queue-time="0ms" rc=1 (error)
      * (19) stop: last-rc-change="Mon Oct 12 14:40:02 2020" last-run="Mon Oct 12 14:40:02 2020" exec-time="6180ms" queue-time="0ms" rc=0 (ok)
```

Where the actual op-history part is the inner-most list (the lines that start with a number in parens).

kgaillot replied:
Man, they're both ugly. :) I guess let's just keep them separate for now. We can try to make them more user-friendly in the future.

clumens replied:
Hah, yeah, they're not great.  I suppose we could look at trying to increase user-friendliness in general for 2.1.0, though it sounds like that is going to be a busy release anyway.  There may not be time.

I would love to see crm_mon, crm_resource, and crm_simulate use the same output functions as much as possible.

kgaillot replied:
Yep, that would be a good goal for the future

kgaillot commented on tools/crm_resource_print.c:
```diff
@@ -33,19 +33,18 @@ cli_resource_print_cts_constraints(pcmk__output_t *out, pe_working_set_t * data_
+                  cons_string(crm_element_value(xml_obj, XML_COLOC_ATTR_TARGET)),
+                  cons_string(crm_element_value(xml_obj, XML_RULE_ATTR_SCORE)),
+                  cons_string(crm_element_value(xml_obj, XML_COLOC_ATTR_SOURCE_ROLE)),
+                  cons_string(crm_element_value(xml_obj, XML_COLOC_ATTR_TARGET_ROLE)));
```
Actually now that tools are gaining xml output it makes sense for cts to use and parse that. But that should be its own project (one of the "would be nice" ones that are always done "next year" ...)

clumens replied:
I could just drop this patch and make doing that conversion the next thing I work on.

kgaillot replied:
No, this is better for now. Maybe at some point, if it becomes a problem or we want a new capability.

kgaillot commented on tools/crm_resource_print.c:
```diff
@@ -306,6 +307,287 @@ property_text(pcmk__output_t *out, va_list args) {
+    if (printed) {
+        out->end_list(out);
+        rc = pcmk_rc_ok;
+    }
```
It's a bit of extra work, but I think it would lessen the chance of future divergence if we have a separate function for actually checking the various conditions, that just sets bits in a bitmask, and then the message functions check those bits. That might end up coming handy in other places anyway.

One issue is wanting to print lock_node. Maybe it could be a struct with a bitmask and lock_node.

That model would be harder to follow with resource-why.

I'm not set on that, we could leave that as an open question.

clumens replied:
Fixed.  I don't see a good way of reducing resource-why, though.  So much of that is iterating across lists hat it doesn't seem to lend itself well to the same model.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -1819,8 +1794,8 @@ cli_resource_execute(pcmk__output_t *out, pe_resource_t *rsc,
-            int rc = cli_resource_search(out, rsc, requested_name, data_set);
-            if(rc > 0 && force == FALSE) {
+            GListPtr rscs = cli_resource_search(out, rsc, requested_name, data_set);
+            if(rscs > 0 && force == FALSE) {
```
rscs != NULL ?

clumens replied:
Heh, whoops.

kgaillot reviewed (changes requested):
Wow, hats off. This is some mighty work.

clumens commented on lib/common/output_xml.c:
```diff
@@ -296,14 +293,14 @@ xml_begin_list(pcmk__output_t *out, const char *singular_noun, const char *plura
 
     if (name == NULL) {
-        name = buf;
+        name = g_ascii_strdown(buf, -1);
```
Ugh, that's no good.  I'll do a separate PR just for that, to be merged before this one.

clumens replied:
Fixed this usage.  I don't see any others, though.

kgaillot replied:
I don't know where else I saw g_ascii_strdown(), guess it was a mistake

clumens commented on tools/crm_resource_runtime.c:
```diff
@@ -23,21 +23,21 @@ do_find_resource(pcmk__output_t *out, const char *rsc, pe_resource_t * the_rsc,
                 state = "Master";
             }
-            fprintf(stdout, "resource %s is running on: %s %s\n", rsc, node->details->uname, state);
+            out->info(out, "resource %s is running on: %s %s", rsc, node->details->uname, state);
```
That's correct.

kgaillot replied:
This is needed in XML too.

Looking at cli_resource_search(), we count and return the nodes found, then ignore that. So I think we can stop counting. :) do_find_resource() essentially prints out a list, so I think we can use a list here, although the items should have the resource, node, and state (actually role) separate in XML.

clumens replied:
This is a little hard to follow, but in d999694 I've gotten rid of the counting.  I also think the XML looks the way you want, but you'd want to double check that.

kgaillot replied:
That looks fine

clumens commented on tools/crm_resource.c:
```diff
@@ -1711,10 +1716,19 @@ main(int argc, char **argv)
+            GListPtr all = NULL;
+            all = g_list_prepend(all, strdup("*"));
+            rc = out->message(out, "resource-list", data_set,
+                              pe_print_printf | pe_print_rsconly | pe_print_pending,
```
Fixed, and another one that was hiding in tools/crm_resource_print.c

clumens commented on tools/crm_resource_print.c:
```diff
@@ -235,75 +235,92 @@ cli_resource_print(pcmk__output_t *out, pe_resource_t *rsc,
+        return pcmk_rc_ok;
     } else {
-        CMD_ERR("Attribute '%s' not found for '%s'", attr, rsc->id);
+        out->err(out, "Attribute '%s' not found for '%s'", attr, rsc->id);
```
It's slow going, but one of my long term goals.  I see I missed a couple usages in crm_resource.  After that, it's down to crm_ticket and crm_rule.

clumens commented on lib/lrmd/lrmd_output.c:
```diff
@@ -0,0 +1,178 @@
+    xmlSetProp(node, (pcmkXmlStr) "for", (pcmkXmlStr) agent_spec);
+
+    rc = lrmd_conn->cmds->list_ocf_providers(lrmd_conn, agent_spec, &list);
+    return xml_list(out, list, rc, "provider");
```
I agree on not having side effects in the output messages.  This seems like another thing to audit at some point.

I could do a single set of custom messages, but that would get rid of some of the extras I added.  The alternatives message adds an attribute for the agent it's listing alternatives for, for instance.  The agents message adds standard and provider attributes.  For crm_resource, you could say it doesn't matter because you'd already know these based on the command line arguments.  However, they seemed useful enough to have in an external library where you may want that kind of information.  I can do whatever, though.

kgaillot replied:
Separate custom messages would also be fine, or a single message could take a whole bunch of optional arguments.

clumens commented on tools/crm_resource_print.c:
```diff
@@ -235,75 +235,92 @@ cli_resource_print(pcmk__output_t *out, pe_resource_t *rsc,
+    if (value != NULL) {
+        out->begin_list(out, NULL, NULL, "Attributes");
+        out->list_item(out, attr, "%s", value);
+        out->end_list(out);
```
I thought we might eventually display a list of attributes somewhere else, so it made sense to me to have attributes in a list everywhere, even if it were just a singleton list.  I suppose if I really thought we would want to display a list of attributes, I would do this outside of crm_resource, and it would take a list instead of a hash table, and I'd have to build a list in crm_resource.  If you think that's worth it, I can do it.

Otherwise, I can just change this.  I guess I'm not sure what the output would look like in that case.

kgaillot replied:
I think just the attribute value by itself, like the current output, is fine. It allows scripting like

    myparam=$(crm_resource -g myparam -r myrsc)

at least as long as the scripter isn't too worried about errors :)

clumens replied:
For text output, I'm already just printing it by itself - see attribute_text.  It's only HTML and XML that will get written out as a list of one item.

kgaillot reviewed (changes requested):
Also bump crm feature set for the new arguments

clumens commented on lib/lrmd/lrmd_output.c:
```diff
@@ -0,0 +1,178 @@
+    xmlNodePtr node = pcmk__output_xml_create_parent(out, "providers");
+
+    xmlSetProp(node, (pcmkXmlStr) "standard",
+               agent_spec == NULL ? (pcmkXmlStr) "ocf" : (pcmkXmlStr) agent_spec);
```
Fixed, I think.

kgaillot replied:
I'm not sure how this works ... is agent_spec standard or standard:provider at this point?

clumens commented on lib/pengine/bundle.c:
```diff
@@ -1589,6 +1589,16 @@ pe__bundle_xml(pcmk__output_t *out, va_list args)
+            char *rsc_xml = dump_xml_formatted(rsc->xml);
+            out->output_xml(out, "xml", rsc_xml);
+            free(rsc_xml);
+        }
```
If we're going to have a separate message just for printing out the resource config, then perhaps we don't need f22783292fbfecbf2d8ff62c795160190f6fa2d2.  I only added that patch to support putting this code into the messages.  And then if that's the case, perhaps we should go ahead and revert it immediately.  It adds to pe_print_options (which looks public to me), but it's only been released in -rc1.  I think it would be okay for us to make that change.

kgaillot replied:
That does sound good. We can keep that commit's formatting changes and just drop the new flag

clumens commented on tools/crm_resource_print.c:
```diff
@@ -223,15 +223,18 @@ int
-    fprintf(stdout, "%sxml:\n%s\n", expanded ? "" : "raw ", rsc_xml);
-    free(rsc_xml);
+    out->message(out, "resource-list", data_set, opts, FALSE, TRUE, FALSE,
+                 FALSE, all, rscs, FALSE);
```
All the resource message variants expect to be called in the context of a list.  There's a couple things we could do here:

* Nothing, and leave the patch as is.
* Make list_item not assert if there's no list, and instead it just prints out.  This is probably reasonable for text but I don't like it elsewhere.  Maybe it's not so bad for XML either.
* I guess I can only think of those two things.

kgaillot replied:
Ah, I forgot about that. Either approach would be fine.

clumens replied:
Oh, here's the other option I'd come up with.  I could replace the call to resource-list with:
```
out->begin_list(out, ...);
out->message(out, crm_map_element_name(), ...);
out->end_list(out);
```

That at least gets rid of all the booleans and the complication of calling that much busier message.

clumens replied:
I'm not happy with how this looks:
```
[root@cluster01 ~]# crm_resource -w -r Fencing --output-as xml
<pacemaker-result api-version="2.3" request="crm_resource -w -r Fencing --output-as xml">
  <resources>
    <resource id="Fencing" resource_agent="stonith:fence_xvm" role="Started" active="true" orphaned="false" blocked="false" managed="true" failed="false" failure_ignored="false" nodes_running_on="1">
      <node name="cluster02" id="2" cached="true"/>
    </resource>
    <xml><![CDATA[...
]]></xml>
  </resources>
  <status code="0" message="OK"/>
</pacemaker-result>
```

kgaillot replied:
Why not? It seems reasonable for passing along external XML, and it's the same as stonith_admin --metadata.

clumens replied:
I don't like that the xml tag is at the same level as the resource it is the XML for.  I'd prefer it to be nested, but that would require modifying the resource message.  Alternately, I could change the enclosing <resources> tag.  We know this is only going to be one element long.

clumens replied:
Fixed by changing the enclosing tag.

clumens commented:
Several things left to do, but I wanted to push another intermediate version while I figure them out.

kgaillot commented on tools/crm_resource_runtime.c:
```diff
@@ -134,12 +134,13 @@ find_resource_attr(pcmk__output_t *out, cib_t * the_cib, const char *attr,
         rc = EINVAL;
-        printf("Multiple attributes match name=%s\n", attr_name);
+        out->info(out, "Multiple attributes match name=%s", attr_name);
+        out->spacer(out);
```
Before this PR, the spacer was after the values rather than before

clumens replied:
Fixed.

kgaillot commented on lib/pengine/pe_output.c:
```diff
@@ -1284,6 +1284,109 @@ pe__node_attribute_html(pcmk__output_t *out, va_list args) {
+    if (rsc) {
+        pe_node_t *node = pe__current_node(rsc);
+        const char *target_role = g_hash_table_lookup(rsc->meta, XML_RSC_ATTR_TARGET_ROLE);
+        int opts = pe_print_printf | pe_print_rsconly | pe_print_suppres_nl | pe_print_pending;
```
Don't need pe_print_printf or pe_print_suppres_nl with formatted output

clumens replied:
Fixed.  This is the danger of just moving code over from an old function to a new one without really studying it.

kgaillot commented on tools/crm_resource_print.c:
```diff
@@ -382,6 +382,76 @@ resource_check_xml(pcmk__output_t *out, va_list args) {
+        xmlNode *sub_node = pcmk__output_create_xml_text_node(out, "node", node->details->uname);
+
+        if (!pe_rsc_is_clone(rsc) && rsc->fns->state(rsc, TRUE) == RSC_ROLE_MASTER) {
+            xmlSetProp(sub_node, (pcmkXmlStr) "state", (pcmkXmlStr) "master");
```
Let's s/master/promoted/ for future proofing

clumens replied:
Good idea.  I've also adjusted the schema to match.

kgaillot commented on include/crm/crm.h:
```diff
@@ -51,7 +51,7 @@ extern "C" {
  * >=3.2.0:  DC supports PCMK_LRM_OP_INVALID and PCMK_LRM_OP_NOT_CONNECTED
  */
-#  define CRM_FEATURE_SET		"3.6.2"
+#  define CRM_FEATURE_SET		"3.6.3"
```
To help future us, add something to the commit message about why it's bumping :)

clumens replied:
Done.

kgaillot reviewed (commented):
I think we're getting close :)

The "Waiting for N replies" message is a progress indicator, initially printed without a newline in start_mainloop(), then in controller_event_cb(), a dot is printed without a newline as each reply comes in and "OK" with a newline is printed at the end. To fit that into the new model; we could:
* Add progress start/continue/end methods to pcmk__output_t (ignored in XML and HTML, text as above)
* Change it to multiline output, e.g. "Waiting for N replies" / "... got reply" / "... got reply (done)"
* info() the "Waiting" and a final "done" and drop the progress dots
* something else?

The "WARNING" about the constraint I think we can just info().

clumens commented on tools/crm_resource_print.c:
```diff
@@ -223,12 +223,20 @@ int
 
-    rsc->fns->print(rsc, NULL, opts, stdout);
+    all = g_list_prepend(all, strdup("*"));
+    rscs = g_list_prepend(rscs, strdup(rsc->id));
```
Nope.

clumens commented on tools/crm_resource_print.c:
```diff
@@ -235,75 +235,92 @@ cli_resource_print(pcmk__output_t *out, pe_resource_t *rsc,
-                " returning the default value for %s", rsc->id, crm_str(attr));
-        current = NULL;
+static void
+add_resource_name(pcmk__output_t *out, pe_resource_t *rsc, bool is_xml) {
```
Hm, nope.  I wonder what I was thinking there.

clumens commented:
> The "Waiting for N replies" message is a progress indicator, initially printed without a newline in start_mainloop(), then in controller_event_cb(), a dot is printed without a newline as each reply comes in and "OK" with a newline is printed at the end. To fit that into the new model; we could:
> 
>     * Add progress start/continue/end methods to pcmk__output_t (ignored in XML and HTML, text as above)
> 
>     * Change it to multiline output, e.g. "Waiting for N replies" / "... got reply" / "... got reply (done)"
> 
>     * info() the "Waiting" and a final "done" and drop the progress dots
> 
>     * something else?

Fixed, with a combination of multiline output and using info().  I don't think this information is especially useful in XML or HTML (or other non-interactive formats, for that matter).  However, I don't think I like it.  In the cleanup function in tools/crm_resource.c, cli_resource_delete is called first (which prints the messages about what gets cleaned up), and then the main loop is started (which prints the messages about waiting for replies).  So the output looks like this:

```
Cleaned up dummy on cluster02
Cleaned up dummy on cluster01
Waiting for 2 replies from the controller
... got reply
... got reply (done)
```

And that seems backwards.  Maybe an easy way to deal with this is to change the cleanup messages to something like "Scheduling cleanup of ...".

kgaillot commented:
> So the output looks like this:
> 
> ```
> Cleaned up dummy on cluster02
> Cleaned up dummy on cluster01
> Waiting for 2 replies from the controller
> ... got reply
> ... got reply (done)
> ```
> 
> And that seems backwards. Maybe an easy way to deal with this is to change the cleanup messages to something like "Scheduling cleanup of ...".

"Requesting clean-up of ..." would be most accurate. I think it's unlikely anyone's relying on that particular output to stay the same ...

clumens commented:
> "Requesting clean-up of ..." would be most accurate. I think it's unlikely anyone's relying on that particular output to stay the same ...

This comes up in several other places, plus there's also the possibility of errors being printed out beforehand too.  See cli_cleanup_all for examples.  So, maybe I'm making a problem where one doesn't really exist, and there's nothing to do here.

clumens commented:
Okay, I think it's time to remove the WIP.  The only stuff I haven't tested is the advanced options, but I don't think they do much with formatted output anyway.

clumens commented on xml/api/resources-2.4.rng:
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is it confusing to have xml/resources-* and xml/api/resources-*?

kgaillot replied:
I don't think so. For 2.1.0 I'm thinking of moving xml/*{rng,xsl} to xml/cib/ anyway.

clumens commented:
Hopefully I'm running out of things to update and push.  I added a generic list schema in a patch way at the beginning of the series, and then fixes some formatted output message arg typing throughout.

kgaillot commented:
retest this please

kgaillot commented on include/pcmki/pcmki_output.h:
```diff
@@ -21,6 +21,8 @@ extern pcmk__supported_format_t pcmk__out_formats[];
 void pcmk__out_epilogue(pcmk__output_t *out, xmlNodePtr *xml, int retval);
 
+void pcmki__register_messages(pcmk__output_t *out);
+
```
We're not using pcmki as a function prefix anywhere else, so I'm inclined to stick with pcmk__, but of course there's already a function with that name. But we can rename one or the other.

clumens replied:
Right, that's why I went with pcmki__ for this function.  I'd kind of like to change the name of the new function.  This is a situation where multiple naming schemes just happen to collide.  Internal functions get named pcmk__, but message registering functions get named _register_messages.

kgaillot commented on cts/cli/regression.tools.exp:
```diff
@@ -2464,10 +2464,10 @@ Set 'test-primitive' option: id=test-primitive-meta_attributes-is-managed set=te
 Multiple attributes match name=is-managed
   Value: false 	(id=test-primitive-meta_attributes-is-managed)
   Value: false 	(id=test-clone-meta_attributes-is-managed)
-
```
The spacer jumped up

kgaillot reviewed (commented):
Looking good, let me know when you're happy with it

clumens commented:
Okay, I think this is as good as it's gonna be until we get some wider use of it.

kgaillot commented:
retest this please

kgaillot commented:
One questionable output change:

    # crm_resource -r petulant -A

    ## Before
        lsb-dummy                                                                    (score=INFINITY, id=lsb-dummy-with-group-1)
    * group-1
      : Node node2                                                                 (score=10, id=location-group-1-node2-10)
        promotable-1                                                                 (score=INFINITY, with role=Master, id=group-1-with-promotable-1)

    ## After
    Colocations:
      * lsb-dummy (score=INFINITY, id=lsb-dummy-with-group-1
      * Locations:
    group-1:
      * Locations:
        * Node node2 (score=10, id=location-group-1-node2-10)
    Colocations:
      * promotable-1 (score=INFINITY, with role=Master, id=group-1-with-promotable-1
      * Locations:

The "before" output was pretty unintelligible to begin with, but it is supposed to read as "lsb-dummy is colocated with group-1, group-1 has a location preference for node2, and group-1 is colocated with promotable-1".

In the "after" output:
* Close parenthesis is missing on some lines
* We shouldn't print "Locations:" if there are none
* It isn't obvious what the "Locations:" and "Colocations:" headings apply to

I'm thinking ideally it would look more like:

    * lsb-dummy (score=INFINITY, id=lsb-dummy-with-group-1)
      * Colocated with:
        * group-1:
          * Locations:
            * Node node2 (score=10, id=location-group-1-node2-10)
          * Colocated with:
            * promotable-1 (score=INFINITY, with role=Master, id=group-1-with-promotable-1)

Or maybe something like:

    * lsb-dummy (score=INFINITY, id=lsb-dummy-with-group-1)
      * Colocated with group-1
        * Location: node2 (score=10, id=location-group-1-node2-10)
        * Colocated with promotable-1 (score=INFINITY, with role=Master, id=group-1-with-promotable-1)

(BTW, the help text says "Specify output format as one of: text (default), xml" but "none" is also accepted -- is none intentionally hidden?)

kgaillot commented:
The -a/-A output can be dealt with separately, let's get this in the wild :)
